addons/mail/static/src/js/utils.js

```diff
@@ -45,21 +42,21 @@ function linkify(text, attrs) {
 
 function addLink(node, transformChildren) {
     if (node.nodeType === 3) {  // text node
-        return linkify(node.data);
+        return linkify(_.escape(node.data));
```

Comment:
Won't this escape content that was not originally escaped?
Eg. `&` would become `&amp;`.
Surely that would be a problem if the `&` is part of the url?

Reply:
`$(`<div id="x">&amp;</div>`)[0].firstChild.data` give you `&`, getting the nodeValue or data of a text fields unescapes it

if & is part of an url this should not be a problem and better to have `&amp;` in an URL.

Thanks to this an URL containing:

`&a=b&amp;c=d`

is no longer broken to:

`&a=b&c=d`

but becomes:

`&amp;a=b&amp;amp;c=d`

which is correct escaping.

so doing `_.escape(node.data)` gives you back the original escaping that was removed (hence the need for this fix):

https://github.com/odoo/odoo/blob/7938a1a58e714dc8c6cae4b384fb5d5a7e231931/addons/mail/static/src/js/composers/basic_composer.js#L364

Reply:
I understand the potential issue, but I don't agree with the current fix.
It is not correct escaping to remove `&` from a link that had it in the first place. It is ok for the text node of the link, but not for the href attribute, as this would break the link, in particular for get parameters.

Reply:
I don't understand what you mean, the correct href for a link `http://google.com/search?q=test&hl=fr` is `http://google.com/search?q=test&amp;hl=fr` so to me it seems more correct now than before.

Reply:
But this escape should be taken care of as part of `linkify` itself when actually building link code from link text, not before the link detection itself that should normally work based on (non html) text.

Reply:
thanks

for & I tested that it wouldn't change anything but I think you are right for other special characters, for example:

```py
http://hello.com/world"!"
```

previous code would detect link `http://hello.com/world` current code detects `http://hello.com/world"` so I will check how this can be improved

Reply:
Good find, I would suggest you take the opportunity to also write a test for this case, and any other that you felt problematic in a different way than anticipated.